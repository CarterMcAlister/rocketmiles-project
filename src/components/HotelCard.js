import React from 'react';
import styled from 'styled-components';

import { colors, breakpoints } from '../styleVariables';
import placeholderImage from '../images/hotel_placeholder.jpg';

const HotelCard = ({ hotel }) => (
    <HotelCardStyled>
        <HotelImage className="image" image={hotel.hotelStaticContent.mainImage.url} />
        <HotelDetails>
            <div className="hotel-name">{hotel.hotelStaticContent.name}</div>
            <div className="location">üìç {hotel.hotelStaticContent.neighborhoodName}</div>
        </HotelDetails>
        <PriceDetails>
            <div>
                <div className="price">
                    <span
                        dangerouslySetInnerHTML={{
                            __html: hotel.lowestAveragePrice.symbol
                        }}
                    />
                    {hotel.lowestAveragePrice.amount.toLocaleString()}
                </div>
                <div className="price-subtitle">per night</div>
            </div>

            <RewardDetails>
                <div className="earn">Earn</div>
                <div className="miles-section">
                    <span className="mile-amount">{hotel.rewards.miles.toLocaleString()} </span>
                    <span className="miles">miles</span>
                </div>
            </RewardDetails>
            <SelectButton>Select</SelectButton>
        </PriceDetails>
    </HotelCardStyled>
);

const HotelCardStyled = styled.div`
    margin: 20px;
    background: ${colors.white};
    display: grid;
    border-radius: 2px;
    overflow: hidden;
    box-shadow: 0 2px 7px -2px rgba(0, 0, 0, 0.12);

    @media ${breakpoints.medium} {
        grid-template-columns: 210px auto 180px;
    }

    :hover {
        box-shadow: 0 2px 7px -2px rgba(0, 0, 0, 0.22);
    }
`;

const HotelImage = styled.div`
    background-image: ${({ image }) => `url(${image}), url(${placeholderImage})`};
    height: 180px;
    background-size: cover;
    min-width: 180px;
`;

const HotelDetails = styled.div`
    padding: 20px 15px 0;

    .hotel-name {
        font-size: 18px;
        font-weight: 500;
        margin-bottom: 6px;

        color: ${colors.black};
    }

    .location {
        color: ${colors.gray};
    }
`;

const PriceDetails = styled.div`
    display: flex;

    padding: 15px;
    justify-content: space-around;

    @media ${breakpoints.medium} {
        flex-direction: column;
    }

    .price {
        color: ${colors.black};
        font-size: 18px;
        font-weight: 500;
    }

    .price-subtitle {
        font-size: 11px;
        color: ${colors.gray};
        font-weight: 500;
    }
`;

const RewardDetails = styled.div`
    .earn {
        font-size: 12px;
        font-weight: 500;
        padding-bottom: 5px;
    }

    .miles-section {
        display: flex;
        align-items: center;
        font-weight: 700;
    }

    .mile-amount {
        color: ${colors.blue};
    }

    .miles {
        font-size: 11px;
        color: ${colors.gray};
        margin-left: 5px;
    }
`;

const SelectButton = styled.button`
    background: ${colors.orange};
    color: ${colors.white};

    :hover {
        opacity: 0.9;
    }
`;

export default HotelCard;
