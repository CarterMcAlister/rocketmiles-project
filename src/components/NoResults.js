import React from 'react';
import styled from 'styled-components';

import { colors } from '../styleVariables';
import noDataImg from '../images/undraw_no_data_qbuo.svg';

const NoResults = ({ searchTerm }) => {
    return (
        <NoResultsContainer>
            <NoResultsImage src={noDataImg} />
            <h1 className="message-title">No matches found for '{searchTerm}'</h1>
            <div className="message">Please try modifying your selection criteria.</div>
        </NoResultsContainer>
    );
};

const NoResultsContainer = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    margin: 30px;
    padding: 20px 10px 60px 10px;

    .message-title {
        font-size: 1.4rem;
    }

    .message {
        padding: 10px 0;
        color: ${colors.gray};
    }
`;

const NoResultsImage = styled.img`
    max-width: 200px;
    padding-bottom: 50px;
`;

export default NoResults;
