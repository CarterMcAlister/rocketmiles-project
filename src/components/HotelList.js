import React, { useState, useEffect } from 'react';
import styled from 'styled-components';

import HotelCard from './HotelCard';
import NoResults from './NoResults';
import Loader from './Loader';

import { colors } from '../styleVariables';

const HotelList = ({ hotels, sortOrder, nameFilter, isLoading }) => {
    const [hotelsFiltered, setHotelsFiltered] = useState([]);

    const sortLowToHigh = (hotelA, hotelB) =>
        hotelA.lowestAveragePrice.amount > hotelB.lowestAveragePrice.amount ? 1 : -1;

    const sortHighToLow = (hotelA, hotelB) =>
        hotelA.lowestAveragePrice.amount < hotelB.lowestAveragePrice.amount ? 1 : -1;

    // Returns filtered list without mutating original object
    const filterAndSort = list => {
        const refinedList = [...list].filter(hotel =>
            hotel.hotelStaticContent.name.toUpperCase().includes(nameFilter.toUpperCase())
        );
        if (sortOrder !== 'recommended') {
            const sortMethod = sortOrder === 'low-to-high' ? sortLowToHigh : sortHighToLow;
            refinedList.sort(sortMethod);
        }

        return refinedList;
    };

    useEffect(() => {
        setHotelsFiltered(filterAndSort(hotels));
    }, [hotels, sortOrder, nameFilter]);

    return (
        <HotelListStyled>
            {isLoading ? (
                <Loader />
            ) : hotelsFiltered && hotelsFiltered.length > 0 ? (
                hotelsFiltered.map(hotel => <HotelCard hotel={hotel} key={hotel.id} />)
            ) : (
                <NoResults searchTerm={nameFilter} />
            )}
        </HotelListStyled>
    );
};

const HotelListStyled = styled.div`
    background: ${colors.offWhite};
    padding: 1px 0;
`;

export default HotelList;
