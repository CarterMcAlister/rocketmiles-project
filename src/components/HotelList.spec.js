import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';

import HotelList from './HotelList';
import HotelCard from './HotelCard';
import NoResults from './NoResults';
import Loader from './Loader';

import { hotelsList } from '../../test/mockData';

describe('HotelList', () => {
    // Mount in act call to properly trigger useEffect and state updates
    let wrapper;
    act(() => {
        wrapper = mount(<HotelList hotels={hotelsList} sortOrder={'recommended'} nameFilter={''} isLoading={false} />);
    });

    it('should display a list of hotel data in the default sort order', () => {
        expect(wrapper.find(HotelCard)).toHaveLength(hotelsList.length);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should sort the hotels by price in ascending order', () => {
        wrapper.setProps({ sortOrder: 'low-to-high' });
        wrapper.update();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should sort the hotels by price in descending order', () => {
        wrapper.setProps({ sortOrder: 'high-to-low' });
        wrapper.update();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should filter the hotels by name', () => {
        wrapper.setProps({ nameFilter: 'omni' });
        wrapper.update();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should show all returned hotels in default order when filters are reset', () => {
        wrapper.setProps({ nameFilter: '', sortOrder: 'recommended' });
        wrapper.update();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should display a loader until results are returned', () => {
        wrapper.setProps({ isLoading: true });
        wrapper.update();
        expect(wrapper.exists(Loader)).toEqual(true);

        wrapper.setProps({ isLoading: false });
        wrapper.update();
        expect(wrapper.exists(Loader)).toEqual(false);
    });

    it('should display a not found page if there are no results for the selected filters', () => {
        wrapper.setProps({ nameFilter: 'no-results' });
        wrapper.update();

        expect(wrapper.exists(NoResults)).toEqual(true);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
