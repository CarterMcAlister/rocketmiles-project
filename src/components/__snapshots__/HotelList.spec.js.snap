// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelList should display a list of hotel data in the default sort order 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH bhhbHv\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Omni Chicago Hotel &amp; Suites Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>579</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">10,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH iDBycH\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Homewood Suites by Hilton Chicago-Lincolnshire</div><div class=\\"location\\">📍 Lincolnshire</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>123</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">12,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH hlJqdV\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Chicago Marriott Downtown Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>234</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">8,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div></div>"`;

exports[`HotelList should display a not found page if there are no results for the selected filters 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bZQynM fEXQKa\\"><img src=\\"test-file-stub\\" class=\\"sc-gzVnrw duVirr\\"><h1 class=\\"message-title\\">No matches found for 'no-results'</h1><div class=\\"message\\">Please try modifying your selection criteria.</div></div></div>"`;

exports[`HotelList should filter the hotels by name 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH bhhbHv\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Omni Chicago Hotel &amp; Suites Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>579</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">10,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div></div>"`;

exports[`HotelList should show all returned hotels in default order when filters are reset 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH bhhbHv\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Omni Chicago Hotel &amp; Suites Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>579</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">10,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH iDBycH\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Homewood Suites by Hilton Chicago-Lincolnshire</div><div class=\\"location\\">📍 Lincolnshire</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>123</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">12,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH hlJqdV\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Chicago Marriott Downtown Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>234</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">8,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div></div>"`;

exports[`HotelList should sort the hotels by price in ascending order 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH iDBycH\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Homewood Suites by Hilton Chicago-Lincolnshire</div><div class=\\"location\\">📍 Lincolnshire</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>123</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">12,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH hlJqdV\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Chicago Marriott Downtown Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>234</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">8,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH bhhbHv\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Omni Chicago Hotel &amp; Suites Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>579</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">10,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div></div>"`;

exports[`HotelList should sort the hotels by price in descending order 1`] = `"<div class=\\"sc-dnqmqq hLiXFM\\"><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH bhhbHv\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Omni Chicago Hotel &amp; Suites Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>579</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">10,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH hlJqdV\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Chicago Marriott Downtown Magnificent Mile</div><div class=\\"location\\">📍 Magnificent Mile</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>234</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">8,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div><div class=\\"sc-bdVaJa jNhWPJ\\"><div class=\\"image sc-bwzfXH iDBycH\\"></div><div class=\\"sc-htpNat fXoYiV\\"><div class=\\"hotel-name\\">Homewood Suites by Hilton Chicago-Lincolnshire</div><div class=\\"location\\">📍 Lincolnshire</div></div><div class=\\"sc-bxivhb kBRbYc\\"><div><div class=\\"price\\"><span>$</span>123</div><div class=\\"price-subtitle\\">per night</div></div><div class=\\"sc-ifAKCX iodTwd\\"><div class=\\"earn\\">Earn</div><div class=\\"miles-section\\"><span class=\\"mile-amount\\">12,000 </span><span class=\\"miles\\">miles</span></div></div><button class=\\"sc-EHOje hVOSSJ\\">Select</button></div></div></div>"`;
